cng/rsa.go

```diff
@@ -325,17 +325,36 @@ func keyVerify(pkey bcrypt.KEY_HANDLE, info unsafe.Pointer, hashed, sig []byte,
 	return bcrypt.VerifySignature(pkey, info, hashed, sig, flags)
 }
 
-func newPSS_PADDING_INFO(h crypto.Hash, saltLen int) (info bcrypt.PSS_PADDING_INFO, err error) {
+func newPSS_PADDING_INFO(pkey bcrypt.KEY_HANDLE, h crypto.Hash, saltLen int, sign bool) (info bcrypt.PSS_PADDING_INFO, err error) {
 	hashID := cryptoHashToID(h)
 	if hashID == "" {
 		return info, errors.New("crypto/rsa: unsupported hash function")
 	}
 	info.AlgId = utf16PtrFromString(hashID)
+
+	// A salt length of -1 and 0 are valid Go sentinel values.
+	if saltLen <= -2 {
+		return info, errors.New("crypto/rsa: PSSOptions.SaltLength cannot be negative")
+	}
+	// CNG does not support salt length special cases like Go do,
+	// so we do a best-effort to resolve them.
 	switch saltLen {
 	case -1: // rsa.PSSSaltLengthEqualsHash
 		info.Salt = uint32(h.Size())
 	case 0: // rsa.PSSSaltLengthAuto
-		err = errors.New("crypto/rsa: rsa.PSSSaltLengthAuto not supported")
+		if sign {
+			// Go sets the salt as large as possible when signing.
+			bits, err := getUint32(bcrypt.HANDLE(pkey), bcrypt.KEY_LENGTH)
+			if err != nil {
+				return info, errors.New("crypto/rsa: key length can't be retrieved" + err.Error())
```

Comment:
Space to keep the inner error readable?

```suggestion
				return info, errors.New("crypto/rsa: key length can't be retrieved " + err.Error())
```